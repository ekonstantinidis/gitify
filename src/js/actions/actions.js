var Reflux = require('reflux');

var Actions = Reflux.createActions({

  'login': {},
  'logout': {},
  'getNotifications': {asyncResult: true},

});

module.exports = Actions;
