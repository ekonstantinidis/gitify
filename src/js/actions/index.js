import { apiRequest, apiRequestAuth } from '../utils/api-requests';

import Constants from '../utils/constants';

function constructGithubUrl(settings) {
  return `https://${settings.get('isEnterprise') ? '' : 'api.'}` +
  `${settings.get('baseUrl')}${settings.get('isEnterprise') ? '/api/v3/' : '/'}`;
}

export function makeAsyncActionSet(actionName) {
  return {
    REQUEST: actionName + '_REQUEST',
    SUCCESS: actionName + '_SUCCESS',
    FAILURE: actionName + '_FAILURE'
  };
}

// Authentication

export const LOGIN = makeAsyncActionSet('LOGIN');
export function loginUser(code) {
  return (dispatch, getState) => {
    const { settings } = getState();
    const url = `https://${settings.get('baseUrl')}/login/oauth/access_token`;
    const method = 'POST';
    const data = {
      'client_id': settings.get('clientId'),
      'client_secret': settings.get('clientSecret'),
      'code': code
    };

    dispatch({type: LOGIN.REQUEST});
    debugger;
    return apiRequest(url, method, data)
      .then(function (response) {
        dispatch({type: LOGIN.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: LOGIN.FAILURE, payload: error.response.data});
      });

  };
};

export const LOGOUT = 'LOGOUT';
export function logout() {
  return { type: LOGOUT };
};


// Notifications

export const NOTIFICATIONS = makeAsyncActionSet('NOTIFICATIONS');
export function fetchNotifications() {
  return (dispatch, getState) => {
    const { settings } = getState();
    // Construct the api!
    const url = `${constructGithubUrl(settings)}notifications?participating=${settings.get('participating')}`;
    const method = 'GET';
    const token = getState().auth.get('token');

    dispatch({type: NOTIFICATIONS.REQUEST});

    return apiRequestAuth(url, method, token)
      .then(function (response) {
        dispatch({type: NOTIFICATIONS.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: NOTIFICATIONS.FAILURE, payload: error.response.data});
      });

  };
};


// Single Notification

export const MARK_NOTIFICATION = makeAsyncActionSet('MARK_NOTIFICATION');
export function markNotification(id) {
  return (dispatch, getState) => {
    const { settings } = getState();
    // Construct the api!
    const url = `${constructGithubUrl(settings)}notifications/threads/${id}`;
    const method = 'PATCH';
    const token = getState().auth.get('token');

    dispatch({type: MARK_NOTIFICATION.REQUEST});

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({type: MARK_NOTIFICATION.SUCCESS, payload: response.data, meta: { id }});
      })
      .catch(function (error) {
        dispatch({type: MARK_NOTIFICATION.FAILURE, payload: error.response.data});
      });
  };
};


// Repo's Notification

export const MARK_REPO_NOTIFICATION = makeAsyncActionSet('MARK_REPO_NOTIFICATION');
export function markRepoNotifications(repoSlug) {
  return (dispatch, getState) => {

    const { settings } = getState();
    // Construct the api!
    const url = `${constructGithubUrl(settings)}repos/${repoSlug}/notifications`;
    const method = 'PUT';
    const token = getState().auth.get('token');

    dispatch({type: MARK_REPO_NOTIFICATION.REQUEST});

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({type: MARK_REPO_NOTIFICATION.SUCCESS, payload: response.data, meta: { repoSlug }});
      })
      .catch(function (error) {
        dispatch({type: MARK_REPO_NOTIFICATION.FAILURE, payload: error.response.data});
      });
  };
};


// Starred

export const HAS_STARRED = makeAsyncActionSet('HAS_STARRED');
export function checkHasStarred() {
  return (dispatch, getState) => {
    const { settings } = getState();
    // Construct the api!
    const url = `${constructGithubUrl(settings)}user/starred/${Constants.REPO_SLUG}`;
    const method = 'GET';
    const token = getState().auth.get('token');

    dispatch({type: HAS_STARRED.REQUEST});

    return apiRequestAuth(url, method, token)
      .then(function (response) {
        dispatch({type: HAS_STARRED.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: HAS_STARRED.FAILURE, payload: error.response.data});
      });

  };
};


// Search

export const SEARCH_NOTIFICATIONS = 'SEARCH_NOTIFICATIONS';
export const CLEAR_SEARCH = 'CLEAR_SEARCH';
export function searchNotifications(query) {
  return {
    type: SEARCH_NOTIFICATIONS,
    query: query
  };
};

export function clearSearch() {
  return {
    type: CLEAR_SEARCH
  };
};


// Settings

export const UPDATE_SETTING = 'UPDATE_SETTING';
export function updateSetting(setting, value) {
  return {
    type: UPDATE_SETTING,
    setting: setting,
    value: value
  };
};


export const TOGGLE_SETTINGS_MODAL = 'TOGGLE_SETTINGS_MODAL';
export function toggleSettingsModal() {
  return {
    type: TOGGLE_SETTINGS_MODAL
  };
};
