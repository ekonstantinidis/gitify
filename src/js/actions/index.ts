import axios from 'axios';
import { parse } from 'url';

import { apiRequest, apiRequestAuth } from '../utils/api-requests';
import {
  getEnterpriseAccountToken,
  generateGitHubAPIUrl,
} from '../utils/helpers';
import Constants from '../utils/constants';
import { LogoutAction, LOGOUT } from '../../types/actions';
import { SettingsState, AppState } from '../../types/reducers';

export function makeAsyncActionSet(actionName) {
  return {
    REQUEST: actionName + '_REQUEST',
    SUCCESS: actionName + '_SUCCESS',
    FAILURE: actionName + '_FAILURE',
  };
}

// Authentication

export const LOGIN = makeAsyncActionSet('LOGIN');
export function loginUser(authOptions, code) {
  const { hostname } = authOptions;
  const isEnterprise = hostname !== Constants.DEFAULT_AUTH_OPTIONS.hostname;

  return (dispatch) => {
    const url = `https://${hostname}/login/oauth/access_token`;
    const method = 'POST';
    const data = {
      client_id: authOptions.clientId,
      client_secret: authOptions.clientSecret,
      code: code,
    };

    dispatch({ type: LOGIN.REQUEST });

    return apiRequest(url, method, data)
      .then(function (response) {
        dispatch({
          type: LOGIN.SUCCESS,
          payload: response.data,
          isEnterprise,
          hostname,
        });
      })
      .catch(function (error) {
        dispatch({ type: LOGIN.FAILURE, payload: error.response.data });
      });
  };
}

export function logout(): LogoutAction {
  return { type: LOGOUT };
}

// Notifications

export const NOTIFICATIONS = makeAsyncActionSet('NOTIFICATIONS');
export function fetchNotifications() {
  return (dispatch, getState: () => AppState) => {
    const method = 'GET';
    const { settings }: { settings: SettingsState } = getState();
    const isGitHubLoggedIn = getState().auth.token !== null;
    const endpointSuffix = `notifications?participating=${settings.participating}`;

    function getGitHubNotifications() {
      if (!isGitHubLoggedIn) {
        return;
      }

      const url = `https://api.${Constants.DEFAULT_AUTH_OPTIONS.hostname}/${endpointSuffix}`;
      const token = getState().auth.token;
      return apiRequestAuth(url, method, token);
    }

    function getEnterpriseNotifications() {
      const enterpriseAccounts = getState().auth.enterpriseAccounts;
      return enterpriseAccounts.map((account) => {
        const hostname = account.hostname;
        const token = account.token;
        const url = `https://${hostname}/api/v3/${endpointSuffix}`;
        return apiRequestAuth(url, method, token);
      });
    }

    dispatch({ type: NOTIFICATIONS.REQUEST });

    return axios
      .all([getGitHubNotifications(), ...getEnterpriseNotifications()])
      .then(
        axios.spread((gitHubNotifications, ...entAccNotifications) => {
          const notifications = entAccNotifications.map(
            (accountNotifications) => {
              const { hostname } = parse(accountNotifications.config.url);

              return {
                hostname,
                notifications: accountNotifications.data,
              };
            }
          );

          const allNotifications = isGitHubLoggedIn
            ? [
                ...notifications,
                {
                  hostname: Constants.DEFAULT_AUTH_OPTIONS.hostname,
                  notifications: gitHubNotifications.data,
                },
              ]
            : [...notifications];

          dispatch({
            type: NOTIFICATIONS.SUCCESS,
            payload: allNotifications,
          });
        })
      )
      .catch((error) =>
        dispatch({ type: NOTIFICATIONS.FAILURE, payload: error.response.data })
      );
  };
}

// Single Notification

export const MARK_NOTIFICATION = makeAsyncActionSet('MARK_NOTIFICATION');
export function markNotification(id, hostname) {
  return (dispatch, getState: () => AppState) => {
    const url = `${generateGitHubAPIUrl(hostname)}notifications/threads/${id}`;
    const method = 'PATCH';

    const isEnterprise = hostname !== Constants.DEFAULT_AUTH_OPTIONS.hostname;
    const entAccounts = getState().auth.enterpriseAccounts;
    const token = isEnterprise
      ? getEnterpriseAccountToken(hostname, entAccounts)
      : getState().auth.token;

    dispatch({ type: MARK_NOTIFICATION.REQUEST });

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({
          type: MARK_NOTIFICATION.SUCCESS,
          meta: { id, hostname },
        });
      })
      .catch(function (error) {
        dispatch({
          type: MARK_NOTIFICATION.FAILURE,
          payload: error.response.data,
        });
      });
  };
}

// Repo's Notification

export const MARK_REPO_NOTIFICATION = makeAsyncActionSet(
  'MARK_REPO_NOTIFICATION'
);
export function markRepoNotifications(repoSlug, hostname) {
  return (dispatch, getState: () => AppState) => {
    const url = `${generateGitHubAPIUrl(
      hostname
    )}repos/${repoSlug}/notifications`;
    const method = 'PUT';

    const isEnterprise = hostname !== Constants.DEFAULT_AUTH_OPTIONS.hostname;
    const entAccounts = getState().auth.enterpriseAccounts;
    const token = isEnterprise
      ? getEnterpriseAccountToken(hostname, entAccounts)
      : getState().auth.token;

    dispatch({ type: MARK_REPO_NOTIFICATION.REQUEST });

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({
          type: MARK_REPO_NOTIFICATION.SUCCESS,
          payload: response.data,
          meta: { hostname, repoSlug },
        });
      })
      .catch(function (error) {
        dispatch({
          type: MARK_REPO_NOTIFICATION.FAILURE,
          payload: error.response.data,
        });
      });
  };
}

// Settings

export const UPDATE_SETTING = 'UPDATE_SETTING';
export function updateSetting(setting, value) {
  return {
    type: UPDATE_SETTING,
    setting: setting,
    value: value,
  };
}
