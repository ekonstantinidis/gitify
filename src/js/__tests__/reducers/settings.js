import reducer from '../../reducers/settings';
import { UPDATE_SETTING, HAS_STARRED } from '../../actions';

describe('reducers/settings.js', () => {
  const initialState = {
    participating: false,
    playSound: true,
    showNotifications: true,
    markOnClick: false,
    openAtStartup: false,
    hasStarred: false
  };

  it('should return the initial state', () => {

    expect(reducer(undefined, {})).to.eql(initialState);

  });

  it('should handle UPDATE_SETTING', () => {

    const actionParticipating = {
      type: UPDATE_SETTING,
      setting: 'participating',
      value: true
    };

    expect(reducer(undefined, actionParticipating)).to.eql({
      ...initialState,
      participating: true
    });

    const actionOpenAtStartUp = {
      type: UPDATE_SETTING,
      setting: 'openAtStartup',
      value: true
    };

    expect(reducer(undefined, actionOpenAtStartUp)).to.eql({
      ...initialState,
      openAtStartup: true
    });

  });

  it('should handle HAS_STARRED.SUCCESS', () => {

    const actionParticipating = {
      type: HAS_STARRED.SUCCESS
    };

    expect(reducer(undefined, actionParticipating)).to.eql({
      ...initialState,
      hasStarred: true
    });

  });

  it('should handle HAS_STARRED.SUCCESS', () => {

    const actionParticipating = {
      type: HAS_STARRED.FAILURE
    };

    expect(reducer(undefined, actionParticipating)).to.eql({
      ...initialState,
      hasStarred: false
    });

  });

});
