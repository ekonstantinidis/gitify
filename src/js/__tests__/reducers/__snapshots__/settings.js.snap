// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/settings.js should handle HAS_STARRED.SUCCESS 1`] = `
Immutable.Map {
  openAtStartup: false,
  hasStarred: true,
  showNotifications: true,
  showSettingsModal: false,
  closeOnClick: false,
  participating: false,
  playSound: true,
  markOnClick: false,
  showAppIcon: "both",
}
`;

exports[`reducers/settings.js should handle HAS_STARRED.SUCCESS 2`] = `
Immutable.Map {
  openAtStartup: false,
  hasStarred: false,
  showNotifications: true,
  showSettingsModal: false,
  closeOnClick: false,
  participating: false,
  playSound: true,
  markOnClick: false,
  showAppIcon: "both",
}
`;

exports[`reducers/settings.js should handle TOGGLE_SETTINGS_MODAL 1`] = `
Immutable.Map {
  openAtStartup: false,
  hasStarred: false,
  showNotifications: true,
  showSettingsModal: true,
  closeOnClick: false,
  participating: false,
  playSound: true,
  markOnClick: false,
  showAppIcon: "both",
}
`;

exports[`reducers/settings.js should handle UPDATE_SETTING 1`] = `
Immutable.Map {
  openAtStartup: false,
  hasStarred: false,
  showNotifications: true,
  showSettingsModal: false,
  closeOnClick: false,
  participating: true,
  playSound: true,
  markOnClick: false,
  showAppIcon: "both",
}
`;

exports[`reducers/settings.js should handle UPDATE_SETTING 2`] = `
Immutable.Map {
  openAtStartup: true,
  hasStarred: false,
  showNotifications: true,
  showSettingsModal: false,
  closeOnClick: false,
  participating: false,
  playSound: true,
  markOnClick: false,
  showAppIcon: "both",
}
`;

exports[`reducers/settings.js should return the initial state 1`] = `
Immutable.Map {
  openAtStartup: false,
  hasStarred: false,
  showNotifications: true,
  showSettingsModal: false,
  closeOnClick: false,
  participating: false,
  playSound: true,
  markOnClick: false,
  showAppIcon: "both",
}
`;
